decomp/semantics/uds.py

```diff
@@ -202,9 +227,18 @@ def add_annotation(self, annotation: 'UDSDataset') -> None:
             the annotations to add to the graphs in the corpus
         """
 
+        # If this is a RawUDSDataset, we need to tell the graph which properties
+        # feature raw annotations, as we need to be able to prevent querying
+        # on raw properties.
+        raw_properties = annotation.raw_properties \
+            if hasattr(annotation, 'raw_properties') else set()
```

Comment:
Could this be an `isinstance` check? If so, is there a reason to use `hasattr`? I recognize that it ensures that the next line will succeed, but what I'm worried about are cases where `NormalizedUDSAnnotation` might have erroneously gotten a `raw_properties` attribute. Are you assuming there may be cases where we have a mix of normalized and raw annotations? If so, the comment above might be changed to specify that, since it talks about the class, not mentioning duck typing.